sql/handler.h

```diff
@@ -4552,7 +4552,7 @@ class handler :public Sql_alloc
   virtual int
   get_parent_foreign_key_list(const THD *thd, List<FOREIGN_KEY_INFO> *f_key_list)
   { return 0; }
-  virtual uint referenced_by_foreign_key() { return 0;}
+  virtual bool referenced_by_foreign_key() { return false;}
```

Comment:
While you are changing this, I would suggest that you also add `const` qualifiers and change also the `int` return type to `void`.

Reply:
Usually, I avoid doing two unrelated things in one commit, and I'm not sure if it's possible to add `const` in all implementations. But if you say so, I'll amend the commit.

Reply:
Only InnoDB supports foreign keys, and both member functions are related to that. Actually, also `handler::get_foreign_key_list()` should return `void`.

More than a decade ago, there was another storage engine [PBXT](https://pbxt.blogspot.com/) that supported `FOREIGN KEY` constraints, but it seems that they did not switch to MariaDB when Oracle changes policies around MySQL.

There is a thin wrapper in Mroonga that would invoke InnoDB, and in fact it does not look like the macro `MRN_SET_WRAP_TABLE_KEY` would be `const` compatible.

Reply:
If the return value of `get_foreign_key_list()` is not needed to signal error conditions, why not just return the `List<>` from it?

Reply:
I was thinking the same, but that collection is actually allocated by the caller, typically from the stack. Only in the rare case that `FOREIGN KEY` constraints exist, some heap memory could be allocated and attached to the object. I assume that there are specific reasons why the custom `List` template is being used instead of something like `std::list` or the C++11 `std::forward_list`.

Reply:
But what is really being allocated for `List`? It's a trivial class that contains just three fields: two pointers and an integer. Nothing is allocated until something gets pushed to the list (except for 24 bytes of stack space, which is always needed, no matter how you pass it).
For the generated machine code, there's no fundamental difference whether you pass it as a reference parameter or as return value - in both cases, the caller allocates stack space. For the coding style, it makes a big difference because now your `List` is truly declared as a return value, and not as some sort of out-parameter kludge.

Reply:
You may very well be right. I usually don’t write functions that would return more than would fit in a single register; `page_id_t` would be an exception on 32-bit targets. This should be easy to check with https://godbolt.org. Feel free to change the return type. There might be some additional overhead in the Mroonga call wrapper, but that storage engine is hardly maintained or used. Even that overhead should be rather negligible, assuming that the `List` is simply copied byte by byte, without any deep-copy constructor.